File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if(s.size() == 0) return 0;
        vector<int>hash(256,-1);
        int st =0,end=0,ans = 1;
        while(end<s.size()){
           if(hash[s[end]] != -1){
              if(hash[s[end]]>=st){
                st = hash[s[end]] + 1;
              }
           }

           ans = max(ans,end-st+1);
           hash[s[end]] = end;
           end++;
        }
        return ans;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
       
       Map<Character, Integer> mp = new HashMap<>();
       int ans = 0,i=0,j=0;
        while(i<s.length()){
          mp.put(s.charAt(i), mp.getOrDefault(s.charAt(i),0)+1);

          while(mp.get(s.charAt(i)) > 1 && i > j){
            mp.put(s.charAt(j), mp.get(s.charAt(j)) - 1);
            if(mp.get(s.charAt(j)) == 0){
                mp.remove(s.charAt(j));
            }
            j++;
          }
          ans = Math.max(ans, (i-j+1));
          i++;
        }

    return ans;
    }
}